update: Prepare facilities for model comparisons

mrshu opened the pull request:
- Create model_utils.py with multiple util functions necessary for model
  comparisons.
- Add models.py with factories that are capable of generating models
  with specific parameters.
- Update model_mlp.py to use these new facilities.

Signed-off-by: mr.Shu mr@shu.io

mrshu committed:
update: Prepare facilities for model comparisons

* Create model_utils.py with multiple util functions necessary for model
  comparisons.

* Add models.py with factories that are capable of generating models
  with specific parameters.

* Update model_mlp.py to use these new facilities.

Signed-off-by: mr.Shu <mr@shu.io>

vidriduch commented on model_utils.py:
```diff
@@ -0,0 +1,58 @@
+    for datum in dataset:
+        img = datum['img']
+        mask = datum['mask']
+        for selection, mask_selection in prepare_pixelized_image(img, mask):
```
@mrshu I believe there should be `window_x` and `window_y` passed here to `prepare_pixelized_image` function or be removed from head of this function since it is not used anywhere

mrshu replied:
@Foxiy just fixed that, thanks!

mrshu committed:
update: Vectorize image spliting functions

* model_utils.py - rewriten function prepare_pixelized_image and
                   prepare_pixelized_dataset to vectorize numpy

Created-by: Foxiy <o.jariabka@gmail.com>
Signed-off-by: mr.Shu <mr@shu.io>

vidriduch commented:
@mrshu This looks good for now. But maybe it would be nice to restructure `model_mlp.py` to by like solver in caffe (http://caffe.berkeleyvision.org/tutorial/solver.html) in the sense it would do basic bookkeeping between models so we can evaluate and compare multiple models more easily. As is written in official documentation of caffe solver:
- (periodically) evaluates the test networks
- snapshots the model and solver state throughout the optimization
- updates the solver state according to learning rate, history, and method

So it would be an object that would hold best model with his history and best hyper-parameters.
It would also had callback to function that would evaluate train and validation accuracy as shown https://keras.io/callbacks/ in example section or similar callbacks to this based on experiment. At the end of testing all the models you could just ask the solver object for some dictionary of losses of all models or ask for best model with best hyper-parameters and visualize it or something. I would love to hear you thoughts on this.

vidriduch committed:
Merge pull request #5 from NaiveNeuron/Foxiy/mergable-vectorize-img-split

update: Vectorize image spliting functions

mrshu committed:
update: Add categorical mlp-based models

* Add model_mlp_categorical.py with mlp-based categorical models.

* Update models.py and model_utils.py with necessary changes.

Signed-off-by: mr.Shu <mr@shu.io>

vidriduch commented on models.py:
```diff
@@ -0,0 +1,28 @@
+from keras.models import Sequential
+from keras.layers.core import Dense, Dropout, Activation
+from keras.optimizers import SGD, Adam, RMSprop
```
@mrshu Just very small nitpick... some of these imports are never used anywhere are we planing on using them later or can they be removed?

mrshu replied:
@Foxiy I am not really sure at this point but will clean it up for now.

mrshu committed:
update: Further small fixes

* Removed unused imports.

* Fixed data type warning.

Signed-off-by: mr.Shu <mr@shu.io>

vidriduch commented:
@mrshu Ok this very nice. Thank you for all fixes and patience with mine nitpicking. Now lets get it merge.
